app/uk/gov/hmrc/eoricommoncomponent/frontend/controllers/registration/GYEHowCanWeIdentifyYouUtrController.scala

```diff
@@ -0,0 +1,91 @@
+import uk.gov.hmrc.eoricommoncomponent.frontend.services.cache.SessionCache
+import uk.gov.hmrc.eoricommoncomponent.frontend.services.registration.MatchingService
+import uk.gov.hmrc.eoricommoncomponent.frontend.views.html.migration._
+import uk.gov.hmrc.http.HeaderCarrier
+
+import scala.concurrent.{ExecutionContext, Future}
+
+@Singleton
+class GYEHowCanWeIdentifyYouUtrController @Inject() (
+  authAction: AuthAction,
+  matchingService: MatchingService,
+  mcc: MessagesControllerComponents,
+  howCanWeIdentifyYouView: how_can_we_identify_you_utr,
+  cdsFrontendDataCache: SessionCache
+)(implicit ec: ExecutionContext)
+    extends CdsController(mcc) {
+
+  def form(service: Service, journey: Journey.Value): Action[AnyContent] =
+    authAction.ggAuthorisedUserWithEnrolmentsAction { implicit request => _: LoggedInUserWithEnrolments =>
+      Future.successful(Ok(howCanWeIdentifyYouView(subscriptionUtrForm, isInReviewMode = false, service, journey)))
+    }
+
+  def submit(service: Service, journey: Journey.Value): Action[AnyContent] =
+    authAction.ggAuthorisedUserWithEnrolmentsAction { implicit request => loggedInUser: LoggedInUserWithEnrolments =>
+      subscriptionUtrForm.bindFromRequest.fold(
+        formWithErrors =>
+          Future.successful(
+            BadRequest(howCanWeIdentifyYouView(formWithErrors, isInReviewMode = false, service, journey))
+          ),
+        formData =>
+          matchOnId(formData, InternalId(loggedInUser.internalId)).map {
+            case true =>
+              Redirect(ConfirmContactDetailsController.form(service, journey))
+            case false =>
+              matchNotFoundBadRequest(formData, service, journey)
+          }
+      )
+    }
+
+  private def matchOnId(formData: IdMatchModel, internalId: InternalId)(implicit hc: HeaderCarrier): Future[Boolean] =
+    retrieveNameDobFromCache().flatMap(ind => matchingService.matchIndividualWithId(Utr(formData.id), ind, internalId))
+
+  private def matchNotFoundBadRequest(individualFormData: IdMatchModel, service: Service, journey: Journey.Value)(
+    implicit request: Request[AnyContent]
+  ): Result = {
+    val errorForm = subscriptionUtrForm
+      .withGlobalError(Messages("cds.matching-error.individual-not-found"))
+      .fill(individualFormData)
+    BadRequest(howCanWeIdentifyYouView(errorForm, isInReviewMode = false, service, journey))
+  }
+
+  // TODO Get rid of `.get`. Now if there is no information Exception will be thrown, understand what should happen if this is not provided
+  private def retrieveNameDobFromCache()(implicit hc: HeaderCarrier): Future[Individual] =
+    cdsFrontendDataCache.subscriptionDetails.map(_.nameDobDetails.get).map { nameDobDetails =>
+      Individual.withLocalDate(
+        firstName = nameDobDetails.firstName,
+        middleName = None,
+        lastName = nameDobDetails.lastName,
+        dateOfBirth = nameDobDetails.dateOfBirth
+      )
+    }
```

Comment:
I think instead of `.get` I'll use `map` on Option and this method can return `Future[Option[Individual]]`. The only place where you're using this is `matchOnId` where you're returning `Boolean`. If there is no information, the new method will return `Future(None)` which you can handle in `matchOnId` method. Map on option which will check the condition and `getOrElse` at the end to have Boolean as return option.

I'll paste those two methods here in few mins. Just doing it.

Reply:
One of the methods

```
private def retrieveNameDobFromCache2()(implicit hc: HeaderCarrier): Future[Option[Individual]] =
    cdsFrontendDataCache.subscriptionDetails.map { details =>
      details.nameDobDetails.map { nameDobDetails =>
        Individual.withLocalDate(
          firstName = nameDobDetails.firstName,
          middleName = None,
          lastName = nameDobDetails.lastName,
          dateOfBirth = nameDobDetails.dateOfBirth
        )
      }
    }
```

Reply:
Second one:
```
private def matchOnId(formData: IdMatchModel, internalId: InternalId)(implicit hc: HeaderCarrier): Future[Boolean] =
    retrieveNameDobFromCache().flatMap { maybeInd =>
      maybeInd match {
        case Some(ind) =>
          matchingService.matchIndividualWithId(Utr(formData.id), ind, internalId)
        case None => Future.successful(false)
      }
    }
```

Reply:
Not sure about false, but I think that if there is no data there should be false for `matchOnId` method.

The only thing that I'm not sure, because with those changes we will handle scenario with missing information as something correct. And here I'm confused, because I'm not sure what means if those details are not presented.

Reply:
That wasn't my TODO - it was on the original controller.    The problem is, what to do if somehow the user hasn't entered their name/dob.
I think even the existing behaviour for failing to "match" may be wrong.   We are showing this as a validation error and inviting users to try again.  I think we had this sort of behaviour somewhere else (for organisation matching) and this was changed to have an explicit "your details didn't match" error page.

Reply:
Ah, that's okay, I missed that this TODO was a part of the implementation right now.